mlflow/types/utils.py

```diff
@@ -478,3 +478,46 @@ def _infer_schema_from_type_hint(type_hint, examples=None):
     else:
         _logger.info("Unsupported type hint: %s, skipping schema inference", type_hint)
         return None
+
+
+def _infer_param_schema(parameters: Dict[str, Any]):
+    if not isinstance(parameters, dict):
+        raise MlflowException(
+            f"Expected parameters to be dict, got {type(parameters).__name__}",
+        )
+
+    def _infer_type_and_shape(value):
+        ndim = np.array(value).ndim
+        if ndim == 1:
+            if not isinstance(value, list):
+                raise MlflowException(
+                    f"Expected parameters to be 1D array or scalar, got {type(value).__name__}",
+                    INVALID_PARAMETER_VALUE,
+                )
+            shape = (-1,)
+        elif ndim == 0:
+            shape = None
+        else:
+            raise MlflowException(
+                f"Expected parameters to be 1D array or scalar, got {ndim}D array",
+                INVALID_PARAMETER_VALUE,
+            )
+        value_type = _infer_numpy_dtype(np.array(value).dtype)
+        return value_type, shape
+
+    param_specs = []
+    invalid_params = []
+    for name, value in parameters.items():
+        try:
+            value_type, shape = _infer_type_and_shape(value)
+            param_specs.append(ParamSpec(name=name, type=value_type, default=value, shape=shape))
+        except Exception as e:
+            invalid_params.append((name, value, e))
+
+    if invalid_params:
+        raise MlflowException(
+            f"Failed to infer schema for parameters: {invalid_params}",
+            INVALID_PARAMETER_VALUE,
+        )
```

Comment:
I tried the following code, which I expected to throw an exception:

```
from mlflow.models import infer_signature

s = infer_signature(["k"], params={"foo": {"bar": "pie"}})
```

An exception was indeed thrown, but the contents were a bit confusing / unexpected:

```
Traceback (most recent call last):
  File "log.py", line 4, in <module>
    s = infer_signature(["k"], params={"foo": {"bar": "pie"}})
  File "/Users/corey.zumar/mlflow/mlflow/models/signature.py", line 218, in infer_signature
    params = _infer_param_schema(params) if params else None
  File "/Users/corey.zumar/mlflow/mlflow/types/utils.py", line 518, in _infer_param_schema
    raise MlflowException(
mlflow.exceptions.MlflowException: Failed to infer schema for parameters: [('foo', {'bar': 'pie'}, Exception('Can not infer object without looking at the values, call _map_numpy_array instead.'))]
```

I expected something like "invalid parameter type dict". Can we fix the message and add test case(s) that assert on the exception message content?